src/controllers/wallet/atomic-swap/tx-proposal.controller.js

```diff
@@ -30,86 +32,114 @@ async function buildTxProposal(req, res) {
 
   const network = req.wallet.getNetworkObject();
 
-  const inputs = req.body.inputs || [];
-  const outputs = req.body.outputs || [];
-  const sendTokens = req.body.send_tokens || [];
-  const receiveTokens = req.body.receive_tokens || [];
+  const sendTokens = req.body.send || { tokens: [] };
+  const receiveTokens = req.body.receive || { tokens: [] };
   const changeAddress = req.body.change_address || null;
   const partialTx = req.body.partial_tx || null;
   let markAsSelected = true;
   if (req.body.lock !== undefined) {
     markAsSelected = req.body.lock;
   }
+  const utxos = [];
 
-  if (inputs.length === 0
-    && outputs.length === 0
-    && sendTokens.length === 0
-    && receiveTokens.length === 0
-  ) {
+  if (sendTokens.tokens.length === 0 && receiveTokens.tokens.length === 0) {
     res.status(400).json({ success: false, error: 'Should have at least one operation' });
     return;
   }
 
   const proposal = partialTx
     ? PartialTxProposal.fromPartialTx(partialTx, network) : new PartialTxProposal(network);
 
-  for (const input of inputs) {
+  if (sendTokens.utxos && sendTokens.utxos.length > 0) {
     try {
-      const txData = await new Promise((resolve, reject) => {
-        txApi.getTransaction(input.txId, data => resolve(data))
-          .catch(err => reject(err));
-      });
+      const currentTs = dateFormatter.dateToTimestamp(new Date());
+      for (const utxo of sendTokens.utxos) {
+        const txData = await new Promise((resolve, reject) => {
+          txApi.getTransaction(utxo.txId, data => resolve(data))
+            .catch(err => reject(err));
+        });
 
-      if (!txData.success) {
-        throw new Error(`Utxo for transaction ${input.txId} and input ${input.index} not found`);
-      }
+        if (!txData.success) {
+          throw new Error(`Utxo for transaction ${utxo.txId} and index ${utxo.index} not found`);
+        }
 
-      const txout = txData.tx.outputs[input.index];
-      proposal.addInput(
-        req.wallet,
-        txout.txId,
-        txout.index,
-        txout.value,
-        txout.decoded.address,
-        {
-          markAsSelected,
-          token_data: txout.token_data,
-          token: txout.token,
-        },
-      );
+        let heightLocked = false;
```

Comment:
I think this validation is important enough to have its own helper method on the lib's `HathorWallet` class itself. A method like `isUtxoUsable` could validate its timelock, reward lock height, _mark as locked_ and others on a single validation response object.

This code extraction doesn't need to be done now, as it's currently only used on this `tx-proposal` method, but it's an important piece of business logic.

Reply:
I agree, maybe we should create an issue for this and include another improvement @pedroferreira1 suggested [here](https://github.com/HathorNetwork/hathor-wallet-lib/pull/408#discussion_r960428742)